iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/Devices.java

```diff
@@ -285,27 +310,95 @@ public DeviceTwinName getDeviceTwinNames()
+            throw e;
+        }
+
+        throw new CompletionException(new ResourceNotFoundException(String.format("Unable to get devices " +
+                "with query %s", query)));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public CompletionStage<TwinServiceListModel> getModuleTwinsByQueryAsync(String query, String
+            continuationToken) throws ExternalDependencyException {
+        try {
+            final Pair<HashMap<String, TwinServiceModel>, String> twinsResult = this.getTwinByQueryAsync(
+                    ModulesQueryPrefix,
+                    QueryConditionTranslator.ToQueryString(query),
+                    continuationToken,
+                    MAX_GET_LIST);
+
+            final String newContinuationToken = twinsResult.getRight();
+            final Map<String, TwinServiceModel> twinsMap = twinsResult.getLeft();
+            final List<TwinServiceModel> listOfTwins = new ArrayList<>(twinsMap.values());
+
+            return completedFuture(new TwinServiceListModel(listOfTwins, newContinuationToken));
+        } catch (InvalidInputException e) {
+            final String message = String.format("Unable to get device twin by query: %s", query);
+            log.error(message, e);
+            throw new CompletionException(message, e);
+        } catch (ExternalDependencyException e) {
+            final String message = String.format("Unable to get device twin by query: %s", query);
+            log.error(message, e);
+            throw e;
+        }
+    }
+
+    private Pair<HashMap<String, TwinServiceModel>, String> getTwinByQueryAsync(
+        final String queryPrefix, final String query, String continuationToken, int numberOfResults)
         throws ExternalDependencyException {
-        String fullQuery = query.isEmpty() ? QueryPrefix : String.format("%s where %s", QueryPrefix, query);
-        QueryOptions options = new QueryOptions();
+        final String fullQuery = query.isEmpty() ? queryPrefix : String.format("%s where %s", queryPrefix,
+                query);
+        final QueryOptions options = new QueryOptions();
         if (continuationToken != null && !continuationToken.isEmpty()) {
             options.setContinuationToken(continuationToken);
         }
 
-        HashMap<String, DeviceTwinServiceModel> twins = new HashMap();
+        final HashMap<String, TwinServiceModel> twins = new HashMap<>();
+        String responseContinuationToken = continuationToken;
         try {
             QueryCollection twinQuery = this.deviceTwinClient.queryTwinCollection(fullQuery);
-            while (this.deviceTwinClient.hasNext(twinQuery) && twins.size() < nubmerOfResult) {
+            while (this.deviceTwinClient.hasNext(twinQuery) && twins.size() < numberOfResults) {
                 QueryCollectionResponse<DeviceTwinDevice> response = this.deviceTwinClient.next(twinQuery, options);
+                responseContinuationToken = response.getContinuationToken();
```

Comment:
Might be good to have a debug statement indicating # of devices found etc

Reply:
Added

Reply:
Thanks